### PR TITLE
General improvements in C++ code

### PR DESCRIPTION
**Pull request by paulromano**
This PR makes some progress on the issues identified in #1174. It makes the following changes:

- All global vectors store `unique_ptr` instead of raw pointers
- The `Particle` class has been improved:
  - The `initialize()` method is now a constructor
  - Data members are identified with trailing underscores
  - Within `LocalCoord`, `xyz` and `uvw` were renamed to `r` and `u` and are now of type `Position` and `Direction`
  - To simplify references to `p.coord_[0].r` and `p.coord_[0].u`, I added accessors that refer to these members as `p.r()` and `p.u()`. I think you'll agree this makes the code a lot more readible in places! For coordinates in the lowest coordinate level, there is also `p.r_local()` and `p.u_local()`.
- `Bank` now stores the particle type directly, rather than a static cast to int. This necessitated that I make `Bank` a subtype of `Particle`. `Bank` also uses the `Position` and `Direction` types as well for consistency.

On top of the refactoring, I've also fixed an issue with lost particles that occur when rectangular lattice boundaries are coincident with regular surfaces. My approach is to use both the position and diretion when determining lattice indices (see `RectLattice::get_indices`). On a single fuel pin model where I was using a lattice in the z direction that coincided with the x and y boundaries, I'm now running millions of particles with no lost particles. I should note that the same approach hasn't been implemented for hexagonal lattices yet since it's a bit more complicated. In that case, I rely on the old `r = r + TINY_BIT*u` trick.

### CODE REVIEW

File: src/cross_sections.cpp [hunk 1]
```diff
@@ -312,7 +312,7 @@ read_ce_cross_sections(const std::vector<std::vector<double>>& nuc_temps,
       }
     } // thermal_tables_
 
-    // Finish setting up materials (normalizing densities, etc.)
+    // Finish setting up.material_s (normalizing densities, etc.)
```
Comment (smharper) [inline src/cross_sections.cpp]: Over-eager regex here

File: src/dagmc.cpp [hunk 2]
```diff
@@ -221,7 +221,7 @@ void load_dagmc_geometry()
       c->material_.push_back(MATERIAL_VOID);
     } else {
       if (using_uwuw) {
-        // lookup material in uwuw if the were present
+        // lookup.material_ in uwuw if the were present
```
Comment (smharper) [inline src/dagmc.cpp]: regex

File: src/simulation.cpp [hunk 3]
```diff
@@ -73,7 +73,7 @@ int openmc_simulation_init()
     t->init_results();
   }
 
-  // Set up material nuclide index mapping
+  // Set up.material_ nuclide index mapping
```
Comment (smharper) [inline src/simulation.cpp]: regex

File: src/source.cpp [hunk 4]
```diff
@@ -153,17 +153,14 @@ Bank SourceDistribution::sample() const
...
-    Position r = space_->sample();
-    site.xyz[0] = r.x;
-    site.xyz[1] = r.y;
-    site.xyz[2] = r.z;
+    site.r = space_->sample();
 
     // Now search to see if location exists in geometry
     int32_t cell_index, instance;
-    int err = openmc_find_cell(site.xyz, &cell_index, &instance);
+    int err = openmc_find_cell(&site.r.x, &cell_index, &instance);
```
Comment (smharper) [inline src/source.cpp]: Not a fan of assuming that the struct has the same spacing as an array.  It probably does, but I doubt it's guaranteed by the standards.

Comment (paulromano) [reply src/source.cpp]: Couldn't slip that one pass you, huh? :) I thought about adding a `static_assert` to ensure there's no extraneous padding in the struct, but it seems safer to just convert this to an array before passing so I'll do that.

Comment (smharper) [reply src/source.cpp]: After all those C++ PRs you'll need a lot more than 3,000 diffs to hide ad hoc coding :p

**Pull request merged**